pkg/controller/podautoscaler/horizontal.go

```diff
@@ -520,6 +486,57 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 	return a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
 }
 
+// normalizeDesiredReplicas takes the metrics desired replicas value and normalizes it based on the appropriate conditions (i.e. < maxReplicas, >
+// minReplicas, etc...)
+func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.HorizontalPodAutoscaler, currentReplicas int32, prenormalizedDesiredReplicas int32) (int32, error) {
+	var minReplicas int32
+	if hpa.Spec.MinReplicas != nil {
+		minReplicas = *hpa.Spec.MinReplicas
+	} else {
+		minReplicas = 0
+	}
+
+	desiredReplicas, err := convertDesiredReplicasWithRules(currentReplicas, prenormalizedDesiredReplicas, minReplicas, hpa.Spec.MaxReplicas)
+	if err != nil {
+		return -1, err
+	}
+
+	if desiredReplicas < prenormalizedDesiredReplicas {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooManyReplicas", "the desired replica count is above the maximum number of allowed replicas")
+	} else if desiredReplicas > prenormalizedDesiredReplicas {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooFewReplicas", "the desired replica count is below the minimum number of allowed replicas")
+	} else {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionFalse, "DesiredWithinRange", "the desired replica count is within the acceptable range")
+	}
+
+	return desiredReplicas, nil
+}
+
+func convertDesiredReplicasWithRules(currentReplicas, desiredReplicas, hpaMinReplicas, hpaMaxReplicas int32) (int32, error) {
+	// Do not upscale too much to prevent incorrect rapid increase of the number of master replicas caused by
+	// bogus CPU usage report from heapster/kubelet (like in issue #32304).
+	scaleUpLimit := calculateScaleUpLimit(currentReplicas)
+
+	minimumAllowedReplicas := int32(math.Max(float64(hpaMinReplicas), 1.0))
+	maximumAllowedReplicas := int32(math.Min(float64(scaleUpLimit), float64(hpaMaxReplicas)))
+
+	if maximumAllowedReplicas < minimumAllowedReplicas {
```

Comment:
I'm tempted to just say drop this error, and then drop the entire `error` return value from both new functions.  I'm not convinced it adds anything, and provides a path for accidentally discarding conditions in the future.

Reply:
The only case I see of that actually happening would be if scaleUpLimit < min. In which case returning min (and maybe indicating this in condition message?) sounds reasonable to me.

Reply:
Yup, makes perfect sense. Will change.